vpr/src/place/place.cpp

```diff
@@ -578,7 +696,12 @@ void try_place(const Netlist<>& net_list,
     /* Gets initial cost and loads bounding boxes. */
 
     if (placer_opts.place_algorithm.is_timing_driven()) {
-        costs.bb_cost = comp_bb_cost(NORMAL);
+        if (num_layers == 1) {
+            costs.bb_cost = comp_bb_cost(NORMAL);
```

Comment:
When you make the cost object, it's possible you could make a template wirelength cost object that automatically makes one of two types of bb_cost objects, and then you invoke the same method on it. Worth discussing with Fahri who has done similar things for the router @duck2 . But not necessary if it is hard to do or understand.

Reply:
Probably better to make every bb a cube. I saw a comment about it becoming 20% slower though? No idea why that would happen...

With the current version, I don't think there's a point templating here: there are only 2 cases and the code is organized as non-class fns so it would be a pain to propagate all the templates.

However as it's the case with `is_flat` I think it's best to hoist `cube_bb` in the global context as soon as possible to avoid adding a param to every fn. That would simplify things a lot

Reply:
Thanks, Vaughn and Fahri. I relocated the `cube_bb` to the global context. If I'm able to encapsulate them within a class, I believe templating it would be a cleaner solution than duplicating all the code.

Regarding the performance slowdown with per-layer BB, the issue arises because we are essentially converting all integer variables in cost functions to vectors. These functions reside inside the inner loop of the placement process. I've also profiled the usage of "per-layer" BB for the 2D architecture (which basically means have a vector of size on instead of an integer) and confirmed that the slowdown is caused by the functions responsible for updating the BB.